### ISSUE TITLE
Pickling of STRtree doesn't save all attributes 

### ISSUE THREAD

**Issue by choosehappy** [Toblerity/Shapely#1033]
I make a list of points:
```
points=[Point(x,y) for x,y in zip(embed_x,embed_y)]
``` 

To be able to identify which point (as its linked to a table elsewhere), I add an ID number:

```
for i in range(len(points)):
	points[i].id=i
```

and lastely create a searchtree, and perform a search, this works as expected:

```
searchtree = STRtree(points)
searchtree.query(points[0])[0].id   --- > returns 0
```

I then pickle the tree:
```
pickle.dump(searchtree,open("searchtree.pkl",'wb'))
```
Later on I load the tree:
```
searchtreeunpickled=pickle.load(open("searchtree.pkl",'rb'))
```

which allows me to search the tree successfully, but the id i set is now gone:

```
searchtreeunpickled.query(points[0])[0].id --> AttributeError: 'Point' object has no attribute 'id'
```

I am unsure why the pickling process is not saving the ID number, or if there is another approach i should be using here?

**Comment by choosehappy** [comment]
I suspect this is actually caused by saving the point itself, since pickling a point with an additional attribute set and unpickling it loses the attribute

![image](https://user-images.githubusercontent.com/9681868/99528990-6373bb80-299f-11eb-95b4-12609ae6a3bf.png)

**Comment by snorfalorpagus** [comment]
There are a couple of issues here. To start, there is some documentation supporting the use of attributes on geometries with STRtree like this: https://github.com/Toblerity/Shapely/issues/618#issuecomment-408891713

The first is that pickling a geometry does not pickle any non-standard attributes that have been added to the object.

```
>>> import pickle
>>> from shapely.geometry import Point
>>> p1 = Point(1, 2)
>>> p1.name = "hello"
>>> print(p1.wkt)
POINT (1 2)
>>> print(p1.name)
hello
>>> data = pickle.dumps(p1)
>>> p2 = pickle.loads(data)
>>> print(p2.wkt)
POINT (1 2)
>>> print(p2.name)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Point' object has no attribute 'name'
```

This is something that hasn't been implemented but probably wouldn't be difficult to do. It would break compatibility with any existing pickle files but I don't think that is an issue as they aren't supposed to be long lived.

The other issue is with the STRtree itself. We pickle geometries by converting them to WKB and restoring them. But there isn't an equivalent to this for the STRtree. I don't think we are ever going to be able to pickle this object. We should probably explicitly raise a NotImplementedError.

**Comment by choosehappy** [comment]
> There are a couple of issues here. To start, there is some documentation supporting the use of attributes on geometries with STRtree like this: #618 (comment)

Absolutely, I saw this and that is where the idea for storing the ID came from

It looks like both of us have reached the same conclusion, that the geometries ultimately don't save non-standard attributes. Your insights helped point me to the reason why: https://github.com/Toblerity/Shapely/blob/6aeb8e7932076fc8700a3c36133255957989464c/shapely/geometry/base.py#L216

As you point out, everything is cast as a wkb, which doesn't support those attributes, so I suspect to have this functionality a more general way (or adding on a separate dictionary for non-standard attributes) would need to be added. this could in theory not break backward compatibility but would break forward compatibility.


Regarding the STRtree, not sure why you can't pickle it? In fact, I can say that I have pickled it and loaded it back without an issue : ) 

this is the hack that I came up with, far from idea but appears to work well.

for saving:

```
    searchtree = STRtree(points)
    ids=[g.id for g in searchtree._geoms[:]]
    pickle.dump([searchtree,ids],open(args.searchoutput,'wb')) #--- https://github.com/Toblerity/Shapely/issues/1033
```

and for loading back:

```
    [searchtree,ids] = pickle.load(open(searchdbfile, "rb" ))
    for id, geom in zip(ids, searchtree._geoms):
        geom.id = int(id)  
```
 
I believe, pickle simply iterates over the geoms and uses each of their reduce methods to save/reload them. Is there a limitation to that I'm not seeing?

**Comment by snorfalorpagus** [comment]
I have looked at the code for STRtree more carefully. There is a hack that makes the tree picklable.

https://github.com/Toblerity/Shapely/blob/73626bce8e99b3f251ad874a41d97d2384e734ce/shapely/strtree.py#L84-L97

It doesn't actually try to pickle the tree. It pickles the geometries in the tree (which are stored to ensure a reference is kept for them and they don't get garbage collected), then recreates the tree when unpickling. Not ideal, but works.

In that case it's just the extra attributes on geometries that need fixing.

**Comment by choosehappy** [comment]
Interesting! 

So if i understand, this has two consequences:
1) if the extra attributes on geometries get pickled, then the pickling/unpickling of STRtree would automatically have the additional attributes taken care of.

2) on the flip side, if the tree is actually re-created each unpickle, and the tree also stores all the geoms within it, there is essentially no reason to ever pickle a tree. it would be slightly cheaper in storage to save only the geoms as a list (this would exclude the arguably minimal STRee overhead), and come at no time penalty since the tree would need to be recreated anyway?

I didn't look very closely at the code, but my assumption was that the pickling would save the organization of the tree itself so it wouldn't have to be recreated. in the case when one has e.g., a million points within it, and the creation time is non-trivial, one could do it once and then cache it via pickling but that doesn't seem to be what actually happens :-\

**Comment by Jeremiah-England** [comment]
> 2. on the flip side, if the tree is actually re-created each unpickle, and the tree also stores all the geoms within it, there is essentially no reason to ever pickle a tree. it would be slightly cheaper in storage to save only the geoms as a list (this would exclude the arguably minimal STRee overhead), and come at no time penalty since the tree would need to be recreated anyway?

Apparently you're not the only one who wanted to pickle the STRtree for something like this (see SO question [here](https://stackoverflow.com/q/64455342/9849440)). I wonder if a comment on what pickling the STRtree actually does should go into the documentation?

**Comment by choosehappy** [comment]
Wow, I'm not the only one! 

Yes a brief note may help. Since tree creation is actually quite quick for small numbers of points, most folks may not even realize what the backend behavior is and maybe painfully surprised when scaling up their "fast" code on a few points becomes "slow" code on many objects. 

I had a quick look at  the GEOS library and it seems out of reach to reasonable pickle/unpickle the underlying tree :-\

**Comment by jorisvandenbossche** [comment]
Indeed, some better documentation around this "gotcha"/disclaimer when pickling STRtree would be useful. 

In general, being able to *actually* serialize the tree would be really nice, but that's something that needs to be developed with the GEOS project (and I don't know if this has been discussed before there, or if there are specific reasons this is currently not implemented).

Regarding pickling of extra attributes on the Geometry objects, I am personally hesitant to add any special ability for this. It feels like a potential rabbit hole, and will also make it less performant.

**Comment by choosehappy** [comment]
needless to say i appreciate your reservations : )

perhaps it is sufficient to only pickle a single "extra" attribute such as
"id". if "id" is a string folks can cram additional information in there,
using e.g., json, to their hearts content but would also allow those of us
that need to know which object is which have that available

that said, it would still break the WKB format, so i suspect this is an "in
for a penny, in for a pound" type situation

**Comment by mullenkamp** [comment]
>  2. on the flip side, if the tree is actually re-created each unpickle, and the tree also stores all the geoms within it, there is essentially no reason to ever pickle a tree. it would be slightly cheaper in storage to save only the geoms as a list (this would  exclude the arguably minimal STRee overhead), and come at no time penalty since the tree would need to be recreated anyway?

I also came to the same conclusion recently through my testing. It would be great from a frequent and repetitive querying perspective to be able to save the STRtrees and quickly load them.
For those whom really need this performance for nearest neighbor queries, you can have a look at [Scikit-Learn unsupervised-neighbors](https://scikit-learn.org/stable/modules/neighbors.html#unsupervised-neighbors). I've always had good luck pickling scikit-learn models.

**Comment by caspervdw** [comment]
I created a feature request at GEOS for the dumping/loading of built tree objects: https://trac.osgeo.org/geos/ticket/1130
